# HG changeset patch
# User Rik <rik@octave.org>
# Date 1532630049 25200
#      Thu Jul 26 11:34:09 2018 -0700
# Node ID 4596a80cbc18ffa724f99ba4e18a0855b5765d42
# Parent  3418d3729a51b87e63d74f000e565600560201dd
Correctly read "Name" as empty_value in dlmread (bug #54029)

* dlmread.cc (Fdlmread): Add #include <cmath> for access to isfinite().
Check whether next_char after reading an Inf, NaN, or NA value is another
letter.  If it is then the string does not correctly match Inf, NaN, or NA and
empty_value is returned.  Add BIST test for bug #54029.

diff --git a/libinterp/corefcn/dlmread.cc b/libinterp/corefcn/dlmread.cc
--- a/libinterp/corefcn/dlmread.cc
+++ b/libinterp/corefcn/dlmread.cc
@@ -23,16 +23,17 @@ along with Octave; see the file COPYING.
 
 // Adapted from previous version of dlmread.occ as authored by Kai
 // Habel, but core code has been completely re-written.
 
 #if defined (HAVE_CONFIG_H)
 #  include "config.h"
 #endif
 
+#include <cmath>
 #include <cctype>
 #include <fstream>
 #include <limits>
 
 #include "file-ops.h"
 #include "lo-ieee.h"
 #include "lo-sysdep.h"
 
@@ -452,16 +453,18 @@ such as text, are also replaced by the @
                       if (! iscmplx)
                         {
                           iscmplx = true;
                           cdata = ComplexMatrix (rdata);
                         }
 
                       cdata(i,j++) = Complex (0, x);
                     }
+                  else if (std::isalpha (next_char) && ! std::isfinite (x))
+                    x = empty_value;
                   else
                     {
                       double y = octave_read_double (tmp_stream);
 
                       if (! iscmplx && y != 0.)
                         {
                           iscmplx = true;
                           cdata = ComplexMatrix (rdata);
@@ -642,9 +645,22 @@ such as text, are also replaced by the @
 %!   fwrite (fid, "1,NA,3");
 %!   fclose (fid);
 %!
 %!   assert (dlmread (file), [1, NA, 3]);
 %! unwind_protect_cleanup
 %!   unlink (file);
 %! end_unwind_protect
 
+%!test <*54029>
+%! file = tempname ();
+%! unwind_protect
+%!   fid = fopen (file, "wt");
+%!   fwrite (fid, "NaNe,bNa,Name,c\n1,NaN,3,Inftest\n-Inf,6,NA,8");
+%!   fclose (fid);
+%!
+%!   assert (dlmread (file), [0, 0, 0, 0; 1, NaN, 3, 0; -Inf, 6, NA, 8]);
+%! unwind_protect_cleanup
+%!   unlink (file);
+%! end_unwind_protect
+
+
 */
